### ISSUE TITLE
CI is failing on when Firefox nightly is making the version transition

### ISSUE THREAD

**Issue by calixteman** [mozilla/pdf.js#16526]
```
 > postinstall
> cross-env PUPPETEER_PRODUCT=firefox node node_modules/puppeteer/install.js

ERROR: Failed to set up Firefox Nightly r116.0a1! Set "PUPPETEER_SKIP_DOWNLOAD" env variable to skip download.
Error: Download failed: server returned code 404. URL: https://archive.mozilla.org/pub/firefox/nightly/latest-mozilla-central/firefox-116.0a1.en-US.linux-x86_64.tar.bz2
    at /home/runner/work/pdf.js/pdf.js/node_modules/@puppeteer/browsers/lib/cjs/httpUtil.js:100:31
    at ClientRequest.requestCallback (/home/runner/work/pdf.js/pdf.js/node_modules/@puppeteer/browsers/lib/cjs/httpUtil.js:77:13)
    at Object.onceWrapper (node:events:628:26)
    at ClientRequest.emit (node:events:513:28)
    at HTTPParser.parserOnIncomingClient (node:_http_client:701:27)
    at HTTPParser.parserOnHeadersComplete (node:_http_common:119:17)
    at TLSSocket.socketOnData (node:_http_client:542:22)
    at TLSSocket.emit (node:events:513:28)
    at addChunk (node:internal/streams/readable:324:12)
    at readableAddChunk (node:internal/streams/readable:297:9)
npm ERR! code 1
npm ERR! path /home/runner/work/pdf.js/pdf.js
npm ERR! command failed
npm ERR! command sh -c cross-env PUPPETEER_PRODUCT=firefox node node_modules/puppeteer/install.js

npm ERR! A complete log of this run can be found in:
```

We should fallback on N-1 when N is failing.

@timvandermeij could you have a look please ?

**Comment by timvandermeij** [comment]
I have looked into this and I would say that this is a bug in the Firefox release process. Puppeteer downloads the Firefox release based on the information from https://product-details.mozilla.org/1.0/firefox_versions.json which is expected to only contain valid, published releases. However, when this issue was created, the JSON file already contained the new version number even though that version hadn't actually been published yet until a few hours later. I would argue that the release process should be the other way around: the artifacts should first be in place before advertising that a new version is available for usage, otherwise clients like Puppeteer will try to fetch it while it's not there yet.

Maybe we can somehow skip the download if we detect this so we keep using the previous cached version, but it's not entirely clear to me if the `trimCache` functionality in Puppeteer will actually keep that one or purge it because a newer version exists according to the JSON. Fixing that would require a whole lot of custom code which I'm not really enthusiastic about.

In short, I think we should report this as a bug upstream in Bugzilla somewhere for the Firefox release team so that they can hopefully change the process to only advertise a new version once it's actually in place to avoid clients getting 404s trying to fetch it.

**Comment by majameel** [comment]
I would like to work on this

**Comment by timvandermeij** [comment]
Closing since I don't see this being actionable (or wanted) for fixing on our end. There is already a Puppeteer issue filed for it and it's a Firefox release process bug; Puppeteer is just a consumer of the API that breaks as a result. Moreover, it's quite rare because it only happened once so far ever since April 2020 when we started using it and one has to be quite lucky to kick off the tests in the interval between the API being changed and the release being published.
